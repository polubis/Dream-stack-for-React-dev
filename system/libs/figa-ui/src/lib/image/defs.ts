import type { DetailedHTMLProps, SourceHTMLAttributes } from 'react';

type ImageHTMLElementProps = Omit<
  DetailedHTMLProps<SourceHTMLAttributes<HTMLSourceElement>, HTMLSourceElement>,
  'children'
>;

interface ImageProps extends ImageHTMLElementProps {
  alt: string;
  lazy?: boolean;
  maxWidth?: string;
  maxHeight?: string;
}

export type { ImageProps, ImageHTMLElementProps };
