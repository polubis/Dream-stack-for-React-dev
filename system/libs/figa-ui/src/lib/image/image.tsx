import { Font } from '../font';
import type { ImageProps } from './defs';

import c from 'classnames';

const Image = ({ className, children, caption }: ImageProps) => {
  return (
    <figure className={c('image', className)}>
      {children}
      {caption && (
        <Font className="image-caption" variant="b1">
          {caption}
        </Font>
      )}
    </figure>
  );
};

export { Image };
