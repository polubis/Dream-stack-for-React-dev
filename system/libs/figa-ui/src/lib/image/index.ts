export * from './image';
export * from './defs';
