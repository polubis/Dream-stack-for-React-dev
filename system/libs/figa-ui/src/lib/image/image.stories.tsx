import type { Story, Meta } from '@storybook/react';

import { Image } from './image';
import { Box } from '../box';
import { Font } from '../font';

export default {
  component: Image,
  title: 'Image',
} as Meta;

const Template: Story = () => {
  return (
    <Box padding={[300, 300, 300, 300]} spacing={[300]}>
      <Font variant="h5">400x200</Font>
      <Image
        maxWidth="400px"
        maxHeight="200px"
        alt="My image"
        src="https://img.freepik.com/premium-wektory/dobry-widok-na-gory-grafika-ilustracja-projekt-koszulki-wektor-sztuki_24519-2593.jpg?w=2000"
      />
    </Box>
  );
};

export const Default = Template.bind({});
Default.args = {};
