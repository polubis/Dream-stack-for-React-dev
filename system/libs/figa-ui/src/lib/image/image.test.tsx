import { fireEvent, render, screen } from '@testing-library/react';
import { Image } from './image';

describe('Image can be used when: ', () => {
  const SRC = 'https://localhost:3000/';

  it('[FRAGILE] assigns classes', () => {
    const { container, asFragment } = render(
      <Image className="my-class" alt="My image" src={SRC} />
    );

    const component = container.querySelector('.image');

    expect(component?.className).toContain('image my-class');
    expect(asFragment()).toMatchSnapshot();
  });

  it('assigns alt text', () => {
    render(<Image alt="My image" src={SRC} />);

    screen.getByAltText(/My image/);
  });

  it('allows to pass max width and height', () => {
    render(
      <Image maxHeight="200px" maxWidth="400px" alt="My image" src={SRC} />
    );

    expect(screen.getByAltText(/My image/).style.maxHeight).toBe('200px');
    expect(screen.getByAltText(/My image/).style.maxWidth).toBe('400px');
  });

  it('[FRAGILE] allows to set lazy attribute for image', () => {
    const { asFragment } = render(<Image lazy alt="My image" src={SRC} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('allows to pass custom properties connected with picture element', () => {
    const spy = jest.fn();

    render(<Image alt="My image" src={SRC} onClick={spy} />);

    fireEvent.click(screen.getByAltText(/My image/));

    expect(spy).toHaveBeenCalledTimes(1);
  });
});
