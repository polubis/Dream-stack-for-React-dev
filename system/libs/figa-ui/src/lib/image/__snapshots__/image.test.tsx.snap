// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image can be used when:  [FRAGILE] allows to set lazy attribute for image 1`] = `
<DocumentFragment>
  <picture
    class="image"
  >
    <img
      alt="My image"
      loading="lazy"
      src="https://localhost:3000/"
    />
  </picture>
</DocumentFragment>
`;

exports[`Image can be used when:  [FRAGILE] assigns classes 1`] = `
<DocumentFragment>
  <picture
    class="image my-class"
  >
    <img
      alt="My image"
      loading="eager"
      src="https://localhost:3000/"
    />
  </picture>
</DocumentFragment>
`;
